### PR TITLE
Implement adadelta

### PR DESCRIPTION
**Pull request by davidmascharka**
This implements the adadelta optimizer, as described in

> ADADELTA: AN ADAPTIVE LEARNING RATE METHOD
> Matthew Zeiler

The algorithm is described at the top right of Page 3. This pull request addresses #24. I've verified behavior by training on the spiral example.

### CODE REVIEW

**Comment by davidmascharka** [comment]
@rsokl review me :grimacing:

**Comment by rsokl** [comment]
Would it be possible to have a hard-coded test case against another library's implementation? Just as a baseline unit test.

**Comment by davidmascharka** [comment]
Yeah, I had wanted to do that against PyTorch, which is what I've done with all the operations we've implemented. Do you think tests is the best place to put that or somewhere else, since it's a bit unorthodox.

**Comment by davidmascharka** [comment]
Is there a better way to do that? It's tedious to throw a real "enterprise scale" parameter into a test and not super flexible. This is what I threw together just now:

``` python
import numpy as np
import mygrad as mg
import torch
from mynn.optimizers.adadelta import Adadelta
from mynn.losses.l1_loss import l1_loss

a = np.random.rand(10, 3, 10, 10)
b = np.random.rand(10, 3, 3, 3)

mynn_b = mg.Tensor(b.copy())
torch_b = torch.tensor(b.copy(), requires_grad=True)

y = a.sum(axis=(1, 2, 3)) * 2

torch_optim = torch.optim.Adadelta([torch_b], lr=0)
mynn_optim = Adadelta(mynn_b, rho=0.9)

for _ in range(1000):
    mg_c = mg.nnet.layers.conv_nd(a, mynn_b, stride=1)
    torch_c = torch.nn.functional.conv2d(torch.tensor(a.copy()), torch_b)
    assert np.allclose(mg_c.data, torch_c.detach().numpy())
    
    torch_loss = torch.nn.functional.l1_loss(torch_c.view(a.shape[0], -1).mean(1), 
                                         torch.tensor(y, dtype=torch.float64))
    mynn_loss = l1_loss(mg_c.mean((1, 2, 3)), y)
    assert np.allclose(torch_loss.detach().numpy(), mynn_loss.data)
    
    torch_loss.backward()
    mynn_loss.backward()
    assert np.allclose(mynn_b.grad, torch_b.grad.numpy())
    
    torch_optim.step()
    mynn_optim.step()
    assert np.allclose(mynn_b.data, torch_b.detach().numpy())
    
    torch_optim.zero_grad()
    mynn_loss.null_gradients()
```

All the asserts pass. Is there something specific you had in mind for testing? There's also the [MINST example](https://github.com/davidmascharka/MyNN/blob/master/examples/mnist.ipynb) which does indeed converge under Adadelta.

**Comment by davidmascharka** [comment]
Going to merge and address this as part of #42
